File: 148_SortList.cpp
//95.17%
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
//Try using merge sorting
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        if(!head || !head -> next)
			return head;
		ListNode *slow(head), *fast(head);
		while(fast -> next && fast -> next -> next){
			slow = slow -> next;
			fast = fast -> next -> next;
		}
		ListNode *secondHalf(sortList(slow -> next));
		slow -> next = NULL;
		ListNode *firstHalf(sortList(head));
		ListNode newHead(0), *pre(&newHead);
		while(firstHalf && secondHalf){
			if(firstHalf -> val < secondHalf -> val){
				pre -> next = firstHalf;
				firstHalf = firstHalf -> next;
			}else{
				pre -> next = secondHalf;
				secondHalf = secondHalf -> next;
			}
			pre = pre -> next;
		}
		if(firstHalf)
			pre -> next = firstHalf;
		else if(secondHalf)
			pre -> next = secondHalf;
		return newHead.next;
    }
};

File: 148_SortList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
##23.13%
class Solution(object):
    def sortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head
        slow, fast = head, head
        while fast.next and fast.next.next:
            slow, fast = slow.next, fast.next.next
        secondHalf = self.sortList(slow.next)
        slow.next = None
        firstHalf = self.sortList(head)
        
        newHead = ListNode(0)
        prev = newHead
        while firstHalf and secondHalf:
            if firstHalf.val < secondHalf.val:
                prev.next, firstHalf = firstHalf, firstHalf.next
            else:
                prev.next, secondHalf = secondHalf, secondHalf.next
            prev = prev.next
        if firstHalf:
            prev.next = firstHalf
        elif secondHalf:
            prev.next = secondHalf
        return newHead.next
		
        